core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -501,15 +501,15 @@ class GroupCoordinator(val brokerId: Int,
 
     groupManager.getGroup(groupId) match {
       case None =>
-        responseCallback(Errors.UNKNOWN_MEMBER_ID)
+        responseCallback(Errors.COORDINATOR_NOT_AVAILABLE)
```

Comment:
Hmm.. I'm not sure we should change this. If the group genuinely doesn't exist, we want the member to resend the JoinGroup rather than retrying. Perhaps we can restrict this change to only the `Dead` state.

Reply:
Interesting, how could consumer actually hit this case in production? Is it due to group migration?

Reply:
This case is not related to migration. For example, a consumer may just have a low session timeout and not commit offsets. If it's heartbeat expires, it could well find that the group no longer exists. We would want it to rejoin rather than retry.

Reply:
There's a very subtle difference between 1) a group is dead, and 2) there's no such group in the metadata cache any more, as a group can only be removed from the metadata cache in three ways:

a) a group must first become empty (but it would not be `Dead` yet). And then after the offset expiry time the cleanup thread may finally be able to remove its offsets, and only when `group.is(Empty) && !group.hasOffsets` it will then transit to `Dead`. And then this group will be removed.

b) Due to offset-topic partition leader migration, the group would be "unloaded". In this case it will transit to `Dead` immediately and then be removed.

c) A `delete-group` API request is received and the group is indeed `empty` already, which it will to transit to `Dead` immediately. And later on the same patten of cleanup thread in a) will eventually remove it from cache.

So a group will always first become `Dead` then be removed from the metadata cache.

And therefore if a sync / heartbeat / commit request comes in and it passed the `isCoordinator` check (means it is still the validator for this group.id) but the group cache does not have it, then it must be that the group once existed but has been transit to `Dead` and removed now. In this case I feel letting the consumer client to go the same routine of re-discover and re-join with non-reset member-id is fine @hachikuji @abbccdda  ?